lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -202,37 +202,87 @@ sub list_scheduled_ajax {
     $self->render(json => {data => \@scheduled});
 }
 
-sub stash_job_and_module_list {
-    my ($self) = @_;
+sub _stash_job {
+    my ($self, $args) = @_;
 
-    my $job_id = $self->param('testid') or return;
-    my $job    = $self->app->schema->resultset('Jobs')->search(
-        {
-            id => $job_id
-        },
-        {
-            prefetch => qw(jobs_assets)
-        }
-      )->first
-      or return;
+    return undef unless my $job_id = $self->param('testid');
+    return undef
+      unless my $job = $self->app->schema->resultset('Jobs')->find({id => $job_id}, $args);
+    $self->stash(job => $job);
+    return $job;
+}
+
+sub _stash_job_and_module_list {
+    my ($self, $args) = @_;
 
+    return undef unless my $job = $self->_stash_job($args);
     my $test_modules = read_test_modules($job);
-    $self->stash(job     => $job);
     $self->stash(modlist => ($test_modules ? $test_modules->{modules} : []));
-    return 1;
+    return $job;
 }
 
 sub details {
     my ($self) = @_;
 
-    $self->stash_job_and_module_list or return $self->reply->not_found;
+    $self->_stash_job_and_module_list or return $self->reply->not_found;
     $self->render('test/details');
 }
 
+sub external {
+    my ($self) = @_;
+
+    $self->_stash_job_and_module_list or return $self->reply->not_found;
+    $self->render('test/external');
+}
+
+sub live {
+    my ($self) = @_;
+
+    my $job          = $self->_stash_job or return $self->reply->not_found;
+    my $current_user = $self->current_user;
+    my $worker       = $job->worker;
+    my $worker_vnc   = ($worker ? $worker->host . ':' . (5990 + $worker->instance) : undef);
+    $self->stash(
+        {
+            ws_developer_url         => determine_web_ui_web_socket_url($job->id),
+            ws_status_only_url       => get_ws_status_only_url($job->id),
+            developer_session        => $job->developer_session,
+            is_devel_mode_accessible => $current_user && $current_user->is_operator,
+            current_user_id          => $current_user ? $current_user->id : 'undefined',
+            worker_vnc               => $worker_vnc,
+        });
+    $self->render('test/live');
+}
+
+sub downloads {
+    my ($self) = @_;
+
+    my $job = $self->_stash_job({prefetch => qw(settings jobs_assets)}) or return $self->reply->not_found;
```

Comment:
do we need to get the `settings` while loading `Logs and Assets`?

Reply:
Good question but the answer is unfortunately yes. The function `test_resultfile_list` makes use of the function `settings_hash`.

Reply:
I compare the other code about `{prefetch => qw() }`,  maybe it needs a `[]`. such as `{prefetch => [qw(settings job_assets)] }`.